### PR TITLE
<concepts>: Implement is-default-initializable via placement new

### LINKED ISSUE #334: LWG-3149 `DefaultConstructible` should require default initialization
LWG-3149 `DefaultConstructible` should require default initialization

This is partially implemented; finishing it requires compiler support:

https://github.com/microsoft/STL/blob/8f4c8163775d665d80642044ce27c4bc696127ce/stl/inc/concepts#L118-L120

Also tracked by Microsoft-internal VSO-896348.

### PR DESCRIPTION
**Pull request by CaseyCarter**
Description
===========

`new T` is correctly ill-formed when `T` cannot be default-initialized (except on MSVC thanks to VSO-1084668).

Fixes #334.

[This is a dual of internal MSVC-PR-237440.]

Checklist
=========

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [x] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

File: stl/inc/concepts [hunk 1]
```diff
@@ -108,9 +109,12 @@ concept constructible_from = destructible<_Ty>
     && __is_constructible(_Ty, _ArgTys...);
 
 // CONCEPT default_initializable
-template <class _Ty> // Per P1754R1 and LWG-3149
+template <class _Ty>
 concept default_initializable = constructible_from<_Ty>
-    && requires { _Ty{}; }; // && __is_default_initializable(_Ty); // TRANSITION, VSO-896348
+    && requires {
+        _Ty{};
+        ::new ((void*)nullptr) _Ty; // is-default-initializable<_Ty>
```
Comment (BillyONeal) [inline stl/inc/concepts]: Wait, there are types which pass the first expression but not this one?!

:sigh:

Comment (CaseyCarter) [reply stl/inc/concepts]: Default-init of `const` objects is ill-formed unless those objects have class type and user-provided default constructors (roughly). So e.g. `new const int[42]` is ill-formed but `const int x[42]{}` is not.

Comment (BillyONeal) [reply stl/inc/concepts]: I see. Ow

Comment (StephanTLavavej) [inline stl/inc/concepts]: 🚨 C cast detected 🚨

File: stl/inc/concepts [hunk 2]
```diff
@@ -13,6 +13,7 @@
 #pragma message("The contents of <concepts> are available only with C++20 concepts support.")
 #else // ^^^ !defined(__cpp_lib_concepts) / defined(__cpp_lib_concepts) vvv
 #include <type_traits>
+#include <vcruntime_new.h>
```
Comment (StephanTLavavej) [inline stl/inc/concepts]: Is this `(core)` OK?

File: tests/std/tests/P0898R3_concepts/test.cpp [hunk 3]
```diff
@@ -1487,68 +1487,83 @@ namespace test_constructible_from {
...
-
-    STATIC_ASSERT(!test<NotDefaultConstructible>());
-    STATIC_ASSERT(!test<DeletedDefault>());
-
-    STATIC_ASSERT(!test<PurePublicDestructor>());
-    STATIC_ASSERT(!test<NotDefaultConstructible>());
+    STATIC_ASSERT(default_initializable<int>);
+#if defined(__clang__) || defined(__EDG__) // TRANSITION, VSO-1084668
+    STATIC_ASSERT(!default_initializable<int const>);
+#else // ^^^ non-MSVC / MSVC vvv
```
Comment (StephanTLavavej) [inline tests/std/tests/P0898R3_concepts/test.cpp]: I'd slightly prefer to see "no workaround / workaround" so we know which side to keep.

File: stl/inc/concepts (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/concepts]: WTB `void*{nullptr}`.

Comment (miscco) [reply stl/inc/concepts]: I saw this in libstdc++ implementation and  decided to go for a walk rather than trying to implement it for libc++

Comment (BillyONeal) [reply stl/inc/concepts]: whee woo whee woo

File: stl/inc/concepts (hunk 2 above)
Comment (BillyONeal) [reply stl/inc/concepts]: It looks like it but probably worth building a dummy program to check

Comment (CaseyCarter) [reply stl/inc/concepts]: Grepping vcruntime for `pragma comment` shows that:
```c++
#if defined(_M_IX86) && defined(_CRT_LEGACY_X86_FLT_EXCEPTIONS) && !defined(_M_CEE_PURE)
    #pragma comment(lib, "legacy_x86_flt_exceptions")
#endif
```
in `vcruntime.h` (which `yvals_core.h` includes) is the only lib comment. https://github.com/microsoft/STL/wiki/The-Difference-Between-Core-And-Non-Core-Headers says:

> "core" refers to headers that can be #include'd into translation units without generating import declarations forcing linking with the standard libraries.

Together, I believe the implication is that all vcruntime headers are core headers.

Comment (StephanTLavavej) [reply stl/inc/concepts]: Sounds good.

Comment (CaseyCarter) [reply stl/inc/concepts]: > It looks like it but probably worth building a dummy program to check
```
C:\Users\Casey\Desktop>type repro.cpp
#include <concepts>
#include <vcruntime_new.h>

int main() {}

C:\Users\Casey\Desktop>cl repro.cpp /nologo /std:c++latest /link /nodefaultlib /ENTRY:main
repro.cpp

C:\Users\Casey\Desktop>cl repro.cpp /nologo /std:c++latest /MD /link /nodefaultlib /ENTRY:main
repro.cpp

C:\Users\Casey\Desktop>
```

**Comment by cbezault** [comment]
Sorry, I probably shouldn't have re-run the tests on this PR, didn't realize your internal PR checks expire in 1.5 hours.

**Comment by CaseyCarter** [comment]
> Sorry, I probably shouldn't have re-run the tests on this PR, didn't realize your internal PR checks expire in 1.5 hours.

🤷‍♂ I would have requeued this out of pure paranoia if you hadn't - I don't want to "sneak" something past the CI and wedge the repo.

**Pull request merged**